### ISSUE TITLE
Totally avoid Pod starvation (HOL blocking) or clarify the user expectation on the wiki

### ISSUE THREAD

**Issue by yqwang-ms** [kubernetes/kubernetes#86373]
**What would you like to be added**:

Seems even after #81263 (#81214), the pod starvation (head-of-line blocking) can only be **besteffort alleviated**, i.e. cannot be **totally avoided**.

This is even more worse in this case https://github.com/kubernetes/kubernetes/issues/86373#issuecomment-570186182

We cannot config the backoff duration to be infinite large to totally avoided it, since it will sacrifice other things, such as large scheduling latency (https://github.com/kubernetes/kubernetes/issues/81214#issuecomment-520008883), more cut in line rates (https://github.com/kubernetes/kubernetes/issues/83834), more frequent preemptions (https://github.com/kubernetes/kubernetes/pull/81698), etc.

So, we have to tune a suitable backoff duration in different clusters, which is too time consuming.
And even if we found a suitable backoff duration **initially**, because whether the starvation happens still depends on at least below things which may be varied from time to time:
1. The scheduling performance in the physical clusters which may be impacted by hardware performance (even worse if scheduler extender is also used), such as hardware becomes slower after used for a lot of time, or network congestions, etc.
2. workload patterns, 
3. cluster events frequency, such as pod/node add/del events
4. K8S versions

So we have to **tune it back and forth**, too much cluster operator maintance efforts.
And current design seems unreasonable, because scheduling behavior depends on too many uncontrollable and volatile things. Thus it causes user feels that our scheduling behavior is confusing and differs in different time and environments.

So, K8S default scheduler still cannot **strictly** respect the behavior on the wiki and the word to deprecate FIFO queue:

> As a result, the higher priority Pod may be scheduled sooner than Pods with lower priority if its scheduling requirements are met. If such Pod cannot be scheduled, **scheduler will continue and tries to schedule other lower priority Pods.**
https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/#effect-of-pod-priority-on-scheduling-order

> After the introduction of Pod Priority and Preemption, we added a new scheduling queue which is Priority aware and has various features to provide fairness and to **ensure that a high priority unschedulable pod does not block the head of the queue**. 
https://github.com/kubernetes/kubernetes/issues/76172#issue-429528520

So, do we have any plan to **totally avoid** pod starvation in future? 
(I have proposed one in bottom)

If not, shall we refine the words on the wiki to **clarify the user expectation** for queuing behavior?
So, the starvation (head-of-line blocking), low priority Pod may be blocked even if cluster has free resources for it, will be clarified to be a by design behavior instead of an issue anymore.
And developers/operators who care about starvation (e.s.p. who works for multi-tenant), can also notice the starvation in advance, then tuning backoff or implememt their own queue sort plugin or even implement their own scheduler, instead of until the real issue happen in prod.

**Possible idea for no need tune backoff duration**:
_By this idea , seems we can avoid head-of-line blocking and large scheduling latency (https://github.com/kubernetes/kubernetes/issues/81214#issuecomment-520008883), more cut in line rates (https://github.com/kubernetes/kubernetes/issues/83834), more frequent preemptions (https://github.com/kubernetes/kubernetes/pull/81698), mentioned above._

Goal: 
Every Pod, even low priority Pod, should be **eventually** tried to schedule while being respect the **priority order** to schedule.

Basic Idea:

0. Remove backoff feature.
1. Before any scheduling, snapshot current scheduling view (such as free resouces, etc).
2. Then, scheduling on the snapshot, only dequeue from the priority queue to schedule, and does not enqueue failed pod back in this step, and only update just allocated/assumed resources, and does not update just freed resources. *Note, during this step, we continue to enqueue newly arrived pods as current behavior, to ensure that at the time the new pod is arrived, it will be still scheduled before lower priority pods in the queue.*
3. Until all pods in the priority queue are dequeued (scheduled), then re-snapshot current scheduling view, so all things are updated, then enqueue all pending pods back to the queue, and start over above scheduling round, i.e. step 2.

In this way, we guarantee that: 
Every free resource is first tried by high priority Pod, then low priority Pod, instead of just blocking or skipping the low priority Pod (which may cause starvation), more formally:
`For every free resource, a pod with priority PT arrives at the time TE, will be scheduled (against all these free resource) before, all pods < PT and > TE.`

So we do not need backoff anymore, and the queuing behavior does not depend on scheduling performance anymore, and it is aligned with current wiki, which is more reasonable, nature and intuitive for most users (maybe the best user expectation).

/sig scheduling
@bsalamat and @draveness  PTAL

**Comment by yqwang-ms** [comment]
@bsalamat and @draveness Any comments for this? Thanks!

**Comment by draveness** [comment]
Thanks. First of all, I see your point of opening this discussion on the behavior of the kubernetes scheduler. 

It makes sense to schedule the pods strictly following the priority which could solve the starvation of high priority unschedulable pod which you mentioned. But the unschedulable pod is unlikely to become schedulable and it could also cause a performance penalty for the newer created pod.

Could you elaborate on how to "clarify the user expectation on the wiki", the current statement seems correct to me.

**Comment by yqwang-ms** [comment]
First thanks @draveness to take a look :)

> But the unschedulable pod is unlikely to become schedulable and it could also cause a performance penalty for the newer created pod.

By using another appoarch to schedule we can avoid these cons.
Please check this appoarch in above section **Possible idea for no need tune backoff duration**.


> Could you elaborate on how to "clarify the user expectation on the wiki", the current statement seems correct to me.

Current wiki is:
> As a result, the higher priority Pod may be scheduled sooner than Pods with lower priority if its scheduling requirements are met. If such Pod cannot be scheduled, **scheduler will continue and tries to schedule other lower priority Pods.**
https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/#effect-of-pod-priority-on-scheduling-order

However, in fact, if there are many pending high priority pods (see  #81214), the default scheduler **will NOT** continue and tries to schedule other lower priority Pods (Because high priority pods are requeued back into the heap again, then becomes head blocking).

**Comment by yqwang-ms** [comment]
Also cc @hex108 @Huang-Wei @everpeace  @ahg-g @NickrenREN Any comments for this? Thanks!

**Comment by hex108** [comment]
> if there are many pending high priority pods (see #81214), the default scheduler will NOT continue and tries to schedule other lower priority Pods (Because high priority pods are requeued back into the heap again, then becomes head blocking).

I think it makes sense that high priority pods blocks lower priority pods. How many priority pods are pending for your case? And how often does the problem occur?

As for your proposal, I have some questions:
* After removing backoff feature, how does scheduler schedule pods that have been attempted to be scheduled 3 times and 5 times differently?
* Would you snaphot the priority queue? If not, `Until all pods in the priority queue are dequeued (scheduled)`, there are always new pods coming, the queue will not be empty. If yes, scheduler will not schedule new pods even they have higher priroity.

**Comment by yqwang-ms** [comment]
Thanks @hex108 :)
> How many priority pods are pending for your case?

To simplify the problem, my case is that the whole cluster is used by different teams (multi-tenant, such as TA, TB), and the nodes are already partitioned with different labels which can only used by cooresponding team (such as label LA can only be used by TA, label LB can only be used by TB).

Assume LA has very **tight** resources, and LB has very **sufficient** resources.

If TA has many high priority pods pending on its **tight** LA (~3k pods), it will **also** block TB's low priority pods to schedule on its **sufficient** LB, even if there are sufficient free resource for these TB's low priority pods.

Currently, to avoid cross team starvation problem, we have to run multiple default scheduler instance, each for one team (configure them to different scheduler name), serving as a dedicated queue.

> And how often does the problem occur?

It depends on the above teams' current workloads against its current resource, they runs data process and deep learning training jobs, which may create many pods in one job, and the total job number also depends on team's plan, but they may only have very few resources. So queuing order is important for them.

**Comment by hex108** [comment]
We could set a resource quota for LA and LB, then there will be not so many pending pods, does it work for you?

**Comment by yqwang-ms** [comment]
@hex108 We have investigated resource quota, but it has below cons:
1. resource quota can only describe resource limitation in quantity, it cannot describe such as "the total resource limit for 2 adjacent GPUs under 1 PCI-E Switch". So, it is not easy to work with scheduler extender, custom scheduling framework, or custom scheduler which using extended resource.
2. resource quota need to dynamically adjust according to heatlh hardware resource, for example, if LA initailly has 1000 nodes, but then 999 nodes become unheath which cannot run pods. So there may be still many pending pods if resource quota is not also decreased accordingly.
3. resource quota will directly reject the pod creation request, so generally controller will backoff retry to create pods. So scheduling now relies on controller to do the backoff, which is far from FIFO, and looks weird.

**Comment by yqwang-ms** [comment]
@hex108 for your questions about my proposal:
> After removing backoff feature, how does scheduler schedule pods that have been attempted to be scheduled 3 times and 5 times differently?

For each scheduling round, i.e. my step 2~3, every pending pods will be scheduled exactly once. So, the only difference is the scheduling order, which is ordered by pod priority.
However, for the same priority, we still have freedom to order pods, such first schedule the 3times failed pod, then schedule the 5times failed pod.


> Would you snaphot the priority queue? If not, `Until all pods in the priority queue are dequeued (scheduled)`, there are always new pods coming, the queue will not be empty. If yes, scheduler will not schedule new pods even they have higher priroity.

No, queue is not snapshoted, only cluster resouce view is snapshoted.
It is the same as current behavior, new pods still can enqueue in flight during step 2, and failed pods can only enqueue during step 3, and each time we still pick the highest priority pods from the priority queue.
For your concerns, if there are always new pods coming, the queue will not be empty, and we will always in step 2. But I think it is reasonable, because we only schedule each pod exactly once, no duplicated pod scheduling, so every pod scheduling is worth to do, right? And we also keeps the order.
Even more, I do not think "there are **always** new pods coming", it is hardly to occur in a real prod cluster.

**Comment by yqwang-ms** [comment]
@hex108 I updated the issue to clarify my proposal, thanks!

**Comment by hex108** [comment]
> 1. resource quota can only describe resource limitation in quantity, it cannot describe such as "the total resource limit for 2 adjacent GPUs under 1 PCI-E Switch". So, it is not easy to work with scheduler extender, custom scheduling framework, or custom scheduler which using extended resource.
> 2. resource quota need to dynamically adjust according to heatlh hardware resource, for example, if LA initailly has 1000 nodes, but then 999 nodes become unheath which cannot run pods. So there may be still many pending pods if resource quota is not also decreased accordingly.

These problem still exsits without resource quota. Resource quota only make things better that it makes pending pods less.

> 3. resource quota will directly reject the pod creation request, so generally controller will backoff retry to create pods. So scheduling now relies on controller to do the backoff, which is far from FIFO, and looks weird to let controller to take this scheduling ownership.

Yes, it has the problem, we have not solved this problem well yet.

**Comment by hex108** [comment]
> Even more, I do not think "there are always new pods coming", it is hardly to occur in a real prod cluster.

If there are batch jobs running in the cluster, it is the normal case that there are always new pods coming. And we should deal with them.

**Comment by yqwang-ms** [comment]
> These problem still exsits without resource quota. Resource quota only make things better that it makes pending pods less.

If we run multiple default scheduler instances (configure them to different scheduler name), each for one team, serving as a dedicated queue. We at least can avoid cross team starvation, and the problem 1 and 2 are already resolved by our scheduler extender.

> If there are batch jobs running in the cluster, it is the normal case that there are always new pods coming. And we should take them into consideration.

Yep, essentially it is pod scheduling speed vs pod creation speed. 
From this comment, https://github.com/kubernetes/kubernetes/issues/81214#issuecomment-520008883, let's assume the pod scheduling speed is 100 pods/s.

So, even for batch jobs case, I do not see it is a real case that, there are constantly, 100 pods/s creation to the cluster, it is something like a DDoS...

Anyway, even if it may be real case, as I mentioned:
> But I think it is reasonable, because we **only schedule each pod exactly once during each round**, **no duplicated** pod scheduling (such as current behaivor keeps scheduling the same high priority pods and may forever skip some low priority pods), so every pod scheduling is **worth** to do, and every pod eventually will be scheduled (current behaivor cannot ensure this), right?

**Comment by hex108** [comment]
> If we run multiple default scheduler instances (configure them to different scheduler name), each for one team, serving as a dedicated queue. We at least can avoid cross team starvation, and the problem 1 and 2 are already resolved by our scheduler extender.

If you have resolved it by extender, there should not be these problems after adding resource quota.

> So, even for batch jobs case, if pod scheduling speed < pod creation speed, the scheduler is already out of capacity even for current scheduling behaiour, the right solution should let user to control the creation speed, or using multiple clusters or multiple schedulers.

Suppose that pod scheduling speed == pod creation speed and pods are continously comming, the pod queue will still not be empty.

> But I think it is reasonable, because we only schedule each pod exactly once during each round, no duplicated pod scheduling (such as current behaivor keeps scheduling the same high priority pods and may forever skip some low priority pods), so every pod scheduling is worth to do, and every pod eventually will be scheduled (current behaivor cannot ensure this), right?

I catch your point, it will be better if we could solve the problem :) In your proposal, before each schedule round ends, high priority pods will not get scheduled again even there are enough resources for them. Is it fair for them?

**Comment by yqwang-ms** [comment]
> If you have resolved it by extender, there should not be these problems after adding resource quota.
I mean to avoid avoid cross team starvation, we resolved it by run **multiple default scheduler** instances, instead of leveraging **extender**. Because currently, extender cannot control how to queuing pods, it can only accept the "to be scheduled pods" from default scheduler's. So extender can hardly help to resolve this queuing issue.

> I catch your point, it will be better if we could solve the problem :) In your proposal, before each schedule round ends, high priority pods will not get scheduled again even there are enough resources for them. Is it fair for them?

Thanks for the catching, haha.
I think it is fair for them. Because, every free resource has already been checked by the high priority pod, (note we do not add new free resource during step 2), so schedule it again is generally useless.

BTW, our guarantee is 
`For every free resource, a pod with priority PT arrives at the time TE, will be scheduled (against all these free resource) before, all pods < PT and < TE.`
It is **fair** because it is more reasonable, nature and intuitive for most users (maybe the best user expectation).

**Comment by hex108** [comment]
> note we do not add new free resource during step 2

That is the problem. There might be some resource released, then free resource meets the demand of the high priroity pods, but it could not get scheduled until the schedule round ends. Howerver these resource might be allocated to new pods with same priority.

And there might be other cases we have not considered. I think a proposal or KEP is needed.

**Comment by yqwang-ms** [comment]
Thanks @hex108 for the suggestion :)

Seems for the same priority, we do not need to guarantee the scheduling order (current scheduler also does not guarantee this due to backoff/unschedulable queue, queue flush features, etc).

Anyway, a KEP would be better, while the issue is still worth to raise now, because I want to know, do we really need to totally avoid Pod starvation (HOL blocking), or just clarify the user expectation on the wiki?

If we think HOL blocking is not an issue, a KEP or proposal may not worth to do, right?

**Comment by yqwang-ms** [comment]
> That is the problem. There might be some resource released, then free resource meets the demand of the high priroity pods, but it could not get scheduled until the schedule round ends. Howerver these resource might be allocated to new pods with same priority.

BTW, for current scheduler, it also has this "problem" (although I do not think it is a problem). The high priroity pods may be in unschedulable queue or backoff queue, it could not get scheduled until the backoff ends. Howerver these resource might be allocated to new pods with same priority.

**Comment by hex108** [comment]
> Seems for the same priority, we do not need to guarantee the scheduling order (current scheduler also does not guarantee this due to backoff/unschedulable queue, queue flush features, etc).

But your proposal might make a unusal case to be a common case, we do not know how long will the schedule round be.

For the HOL problem, I have two suggestions to help alleviate it.

* Use reource quota.
* Give TB's pods high priroity.

**Comment by yqwang-ms** [comment]
> we do not know how long will the schedule round be.
This is similar to: currently we also do NOT know the scheduling speed (and other factors mentioned in the issue), so hard to tune backoff duration. So, the same question, why we have to insist the backoff mechanism? I see space to improve it, even though my proposal is not the best one.

Maybe K8S need to support multi-tenant as first class, just using reource quota per namespace and global meaningful priroity, cannot truely isolate the scheduling (such as HOL blocking) for different tenants.

**Comment by hex108** [comment]
> This is similar to: currently we also do NOT know the scheduling speed (and other factors mentioned in the issue), so hard to tune backoff duration. So, the same question, why we have to insist the backoff mechanism? I see space to improve it, even though my proposal may not the best one.

I do not insist the backoff mechanism, I just said my concerns about you proposal, and we need discuss them more :)

As to multi-tenant, it is really a big topic, there is [multi-tenancy](https://github.com/kubernetes-sigs/multi-tenancy), and [kube-batch](https://github.com/kubernetes-sigs/kube-batch) also tries to resolve it.

**Comment by yqwang-ms** [comment]
Agree :) Thank you!
Wish this issue to be a starting point to discuss K8S HOLB issue, and more devs can provide more ideas.

We are also working on the multi-tenant by [HiveD scheduler](https://github.com/microsoft/pai/blob/master/docs/system_architecture.md), so we are truely impacted by this issue.
Seems both scheduler extender (like our HiveD) and scheduling framework are hardly to solve it, and we do not want to rewrite and maintain a whole scheduler like kube-batch.

BTW, I see join the [multi-tenancy](https://github.com/kubernetes-sigs/multi-tenancy) repo need to be a k8s member, would you mind to be my sponsor? :)

Thanks

**Comment by hex108** [comment]
> Wish this issue to be a starting point to discuss K8S HOLB issue, and more devs can provide more ideas.

:)

> BTW, I see join the multi-tenancy repo need to be a k8s member, would you mind to be my sponsor? :)

It requires @kubernetes-sigs member, I do not meet the requirments to be a sponsor, sorry. And it does not need to be a member to use the repo or file issues/PRs.

**Comment by yqwang-ms** [comment]
Thanks @hex108 for the advice :)

**Comment by ahg-g** [comment]
Thanks for bringing up this problem. I think this is something that will come back to us frequently in the future as people start to use k8s in large scale multi-tenant environments. So we should either try to make the scheduler do better, make a recommendation to use multiple schedulers in such setups or make recommendations on how to create workloads to reduce the impact of this problem.

To start the discussion, let me try to summarize the issues that is causing starvation in your case: 
1) the way the scheduler queue is currently managed assumes that the resources in the cluster are shared, but in reality they are not since each tenant is getting a dedicated set of resources (which I think defeats one the main advantages of multi-tenancy). 
2) the scheduler's queue is ordered by priority, and priority is global, but in your case it is actually per tenant 

Regarding your proposal, I think it is very early: it needs to address issues related to how we deal with preemption and cluster autoscaler among other things. It also seems like a solution that overfits the problem you have.

**Comment by yqwang-ms** [comment]
Thanks @ahg-g for comments.
> but in reality they are not since each tenant is getting a dedicated set of resources (which I think defeats one the main advantages of multi-tenancy).

This will not necessarily defeats the multi-tenant pros. Because generally, for a multi-tenant scheduling strategy, user can submit high priority (i.e. guaranteed) workloads on its own dedicated resources, but it can only submit low priority (i.e. overused or opportunistic) workload on other tenant's dedicated resources. So, when other tenant's high priority workloads arrive, its low priority workload can be preempted reasonably.
This strategy preserves multi-tenant pros (sharing resource), while isolate different tenants (safety sharing). This is adopted by some well-known schedulers like HADOOP YARN, Microsoft internal Cosmos, above Hived, etc.


> Regarding your proposal, I think it is very early

My proposal wants to totally fix HOLB inter-tenant and intra-tenant. And if we think we only need to fix inter-tenant HOLB (as keep sharing safety is more important, since inside a tenant, tenant admin can make its own strategy to avoid HOLB, such as kill bocking job manually), multiple queues like YARN, should be another worth approach to try.

Any comment?

**Comment by ahg-g** [comment]
> Thanks @ahg-g for comments.
> 
> > but in reality they are not since each tenant is getting a dedicated set of resources (which I think defeats one the main advantages of multi-tenancy).
> 
> This will not necessarily defeats the multi-tenant pros. Because generally, for a multi-tenant scheduling strategy, user can submit high priority (i.e. guaranteed) workloads on its own dedicated resources, but it can only submit low priority (i.e. overused or opportunistic) workload on other tenant's dedicated resources. So, when other tenant's high priority workloads arrive, its low priority workload can be preempted reasonably.
> This strategy preserves multi-tenant pros (sharing resource), while isolate different tenants (safety sharing). This is adopted by some well-known schedulers like HADOOP YARN, Microsoft internal Cosmos, above Hived, etc.

Right, but in https://github.com/kubernetes/kubernetes/issues/86373#issuecomment-570186182 you say "and the nodes are already partitioned with different labels which can only used by cooresponding team ", which indicates that the teams didn't share any resources. 

> 
> > Regarding your proposal, I think it is very early
> 
> My proposal wants to totally fix HOLB inter-tenant and intra-tenant. And if we think we only need to fix inter-tenant HOLB (as keep sharing safety is more important, since inside a tenant, tenant admin can make its own strategy to avoid HOLB, such as kill bocking job manually), multiple queues like YARN, should be another worth approach to try.
> 
> Any comment?

We don't have the concept of a tenant in k8s (perhaps namespaces is the closest to tenants), and so we need to introduce a new type to represent a queue resource which is a significant effort.

Back to the current implementation, what do you think is the most problematic in your case: the backoff time is not configured properly or that there are too many events triggering false move from the backoff to the active queue? if the latter, do you know which particular events?

**Comment by yqwang-ms** [comment]
> which indicates that the teams didn't share any resources.
See the head line of https://github.com/kubernetes/kubernetes/issues/86373#issuecomment-570186182, I have simplified the problem, so reader can foces on the HOLB. Let me update the comment to include more details.

> what do you think is the most problematic in your case
1. the backoff time need to be tuned back and forth, too much cluster operator maintance efforts.
2. large backoff time will sacrifice other things, such as large scheduling latency (#81214 (comment)), more cut in line rates (#83834), more frequent preemptions (#81698), etc.
3. events may be a potienal trigger, but from my obersevation, it is the every 60s flushUnschedulableQLeftover caused the high priority pods backed to active queue, then lower priority pods never have chance to schedule. Can we also make the 60s flush configurable?

**Comment by ahg-g** [comment]
> events may be a potienal trigger, but from my observation, it is the every 60s flushUnschedulableQLeftover caused most high priority pods backed to active queue, then lower priority pods never have chance to schedule. Can we also make the 60s flush configurable?

We can, feel free to create an issue and send a PR.

**Comment by yqwang-ms** [comment]
Thanks @ahg-g, I can handle the PR later.
BTW, it is also just an alleviator like #81263, we can continue discuss the final solution here :)

**Comment by ahg-g** [comment]
Let me also try to summarize their use case in a few bullet points:
1. They create multi-tenant clusters, tenants are teams in the same org
2. Each tenant gets their own dedicated group of nodes (think node pool)
3. Each tenant can create high or low priority pods
4. High priority pods can only be scheduled on a tenant's dedicated node pool (using taints and tolerations)
5. To improve utilization, low priority pods can be scheduled anywhere, but the preference is in the tenant's node pool (using preferred node affinity)

The scenario where the scheduler could become slow in the above use case: 
1. Tenant A node pool is fully allocated
2. Tenant A creates additional high priority pods, which fail to schedule because the tenant's node pool is allocated
3. Tenant B node pool has plenty of resources
4. Tenant B creates many low priority pods
5. The scheduler is very slow in scheduling Tenant B low priority even though Tenant B's node pool has enough resources. This is because the scheduler keeps wasting scheduling cycles on retrying to schedule Tenant A's high priority pods (remember that pod priority is global)

Solutions to this problem fall under two broad categories (not mutually exclusive):
1) Improve the scheduler's heuristics on when to retry to schedule previously failed pods, or how to order them in the scheduling queue
2) Setup admission webhooks or resource quotas to prevent creating many unschedulable high-priority pods

**Comment by ahg-g** [comment]
I am trying to find out if there are other similar use cases that could inform what solution we go for so that it is generic enough.

**Comment by alculquicondor** [comment]
> Back to the current implementation, what do you think is the most problematic in your case: the backoff time is not configured properly or that there are too many events triggering false move from the backoff to the active queue? if the latter, do you know which particular events?

@yqwang-ms, could you try to explore this question a bit more? I want to understand if (at least for your use case) is the 60s flush what is causing the problem or some other events (like new nodes registering to the cluster)

That said, we probably want a more general solution. I simple solution I can think of is the following:
We introduce a  "meta" priority internal to the scheduler. This priority would include the actual priority, creation timestamp and number of retries, having a weight for each of them.

**Comment by ahg-g** [comment]
> That said, we probably want a more general solution. I simple solution I can think of is the following:
We introduce a "meta" priority internal to the scheduler. This priority would include the actual priority, creation timestamp and number of retries, having a weight for each of them.

Yeah, that is one way to implement queue re-ordering, here is another issue proposing to de-prioritizing pods: https://github.com/kubernetes/kubernetes/issues/85792

**Comment by ahg-g** [comment]
Once https://github.com/kubernetes/kubernetes/issues/87160 is addressed, we can try different sort plugins to test different heuristics.

**Comment by Huang-Wei** [comment]
My 2 cents:

One solution is to add support for multiple-queues, but it's complex and inevitably involves other parts like how to make a strategy to pop pods from different queues, and probably we need to define new API objects. Also, preemption needs to be re-considered.

The other solution I can think of is to leverage some existing algorithm/research to _properly_ (re)place the Pods, e.g., reward-based scheduling - every time a Pod failed to be scheduled, expose a penalty so that its weight is adjusted. With this, we don't rely purely on `Priority` or fixed backOff settings.

**Comment by yqwang-ms** [comment]
@ahg-g and @Huang-Wei 
Thanks for your summary and proposal!
One more thing to add:
Even though this issue is more serious in a multiple tenant cluster, it also impacts other kind of scenarios.

To me, this issue can be splited to 2 different issues to solve (with different work item priority):
1. Avoid **inter-tenant impacts** (including HOLB) caused by queuing, or make a fair queuing policy for all tenants.
2. Avoid **intra-tenant HOLB** (this is to totally fix #81214, even it is a single tenant cluster)

I agree we should solve 1 first, because it is more serious. But we can also discuss on 2, such as is it by design, or we also plan to totally fix it in future?

> Once #87160 is addressed, we can try different sort plugins to test different heuristics.
> The other solution I can think of is to leverage some existing algorithm/research to properly (re)place the Pods, e.g., reward-based scheduling - every time a Pod failed to be scheduled, expose a penalty so that its weight is adjusted. With this, we don't rely purely on Priority or fixed backOff settings.

Theoretically, we can implement a single queue to **simulate** multiple queues, by such as de-prioritizing, or other heuristics re-ordering.
It may be easier to implement it based on current code base, but for reminder, the actual queuing strategy seems do not also become easier. Because you still need to think about:
1. Still need a **clear strategy** defines which tenant's pod should be scheduled first, given multiple tenants's pending pods.
   For example, in YARN, basically, it will choose to schedule the tenant (i.e. YARN queue) which has the lowest usage percentage (i.e. tenant allocated resouce / tenant total guaranteed resource) first.
2. **Preemption** still needs to be re-considered:
    For example, tenant A should only preempt other tenant's lower priority pods, **only** if they are using A's dedicated resource. I.e. Pod priority should only be meaningful inside one tenant. So we at least need an field to let user can specify a Pod belongs to which tenant. (Or shall we reuse namespace as tenant?)
3. The queuing may need to **aware "the total guaranteed resource for each tenant"**, to be clever in heuristics re-ordering (such as in 1). However, we may have to make a new object/API to describe "the total guaranteed resource for each tenant", so that user can configure them.
4. User may still want to aware current effective priority in scheduler (i.e. the internal priority you said), so that they can take actions, such as migrate to another free cluster, if it is de-prioritized to a very low one in current cluster.
5. etc


Anyway, I agree that your proposal seems to be easier to implement, which can be a short term solution. For whether it can also be a final long term solution, we still need to discuss, right? :)

**Comment by yqwang-ms** [comment]
@alculquicondor For my case, it is the 60s flush, but I am not sure if we increase the 60s to longer enough can solve my case, so I used **multiple default scheulder** to serve as multiple queues to solve the **1. inter-tenant impacts**.

Check this log, we did not see MoveAllToActiveQueue (event trigger), but we see frequent movePodsToActiveQueue (flush trigger):
```
I1218 08:30:34.547806       1 scheduling_queue.go:512] BOTEST: flushUnschedulableQLeftover count 543
I1218 08:31:04.553259       1 scheduling_queue.go:512] BOTEST: flushUnschedulableQLeftover count 540
I1218 08:31:34.558404       1 scheduling_queue.go:512] BOTEST: flushUnschedulableQLeftover count 539
I1218 08:32:04.563195       1 scheduling_queue.go:512] BOTEST: flushUnschedulableQLeftover count 548
I1218 08:32:34.568563       1 scheduling_queue.go:512] BOTEST: flushUnschedulableQLeftover count 562
I1218 08:33:04.573045       1 scheduling_queue.go:512] BOTEST: flushUnschedulableQLeftover count 560
I1218 08:33:34.578893       1 scheduling_queue.go:512] BOTEST: flushUnschedulableQLeftover count 559
```

**Comment by ahg-g** [comment]
> Anyway, I agree that your proposal seems to be easier to implement, which can be a short term solution. For whether it can also be a final long term solution, we still need to discuss, right? :)

@yqwang-ms to clarify, I am not saying that tinkering with the queue order will address complex multi-tenant scheduling problems. The default scheduler is by design simple, but at the same time fits the needs of a significant number of use cases. 

In my opinion, re-implementing the YARN scheduler is a significant task that I am not sure it can reasonably be done by refactoring the default scheduler. What is nice about k8s is that the scheduler is just another controller, and k8s makes it very simple to run another scheduler in the cluster beside the default one. We have sig sponsored projects that serve as incubators for such significant efforts.

**Comment by alculquicondor** [comment]
About short term solution:
Sounds like we are in the same line about using an internal priority that penalizes scheduling retries. As stated earlier, this should be easy to do after #87160. @Huang-Wei, do you know of any specific papers for this problem?

About long term solution:
The problem about multi-tenancy is that it is not very well defined in k8s terms. We can choose to use namespaces, but there is a tenant definition being incubated as well: https://github.com/kubernetes-sigs/multi-tenancy/tree/master/tenant. Given this, we could incubate a "multi-tenant scheduler". But what we can do in kube-scheduler is to make the prototyping of such scheduler easier by enhancing the scheduler framework, if needed.

@yqwang-ms you have mentioned preemption, which is a totally separate discussion from starvation and I think it's worth its own issue in the context of multi-tenancy.

**Comment by yqwang-ms** [comment]
@ahg-g Thanks for the clarification, seems we do not plan to **first class** support multi-tenant in default scheduler at least in short term.

Anyway, as @alculquicondor said, even in short term, we at least need to make the default scheduler 
 (and the whole K8S) **more extensible**, so that scheduling plugin and/or scheduler extender is **able to easily** support multi-tenant by themselves. (This kind of prototyping may also make the **first class** support possible and smooth)

For the **more extensible**, I mean we may also need to support more features, such as:
1. Queue sort plugin #87160 
2. Requeue plugin
It is used to judge whether a failed scheduled pod should be requeued now. So that plugin can override the behaiour of unschedulable queue, backoff queue, etc which also impacts on this issue.
For example, it should provide a func ShouldRequeueNow(pod) bool.
3. Queue empty listener plugin
When the active queue is empty, it will call this plugin. This helps to implement the behaiour in my first proposal, i.e. only after it is called, we can requeue the failed scheduled pods, and resnapshot the scheduling view
4. Preemption plugin
It is a must for multi-tenant as alculquicondor said, but could not find it in https://github.com/kubernetes/enhancements/blob/master/keps/sig-scheduling/20180409-scheduling-framework.md
5. Object to define the tenant deicated resource
Maybe we can reuse the namespace and resource quota object.

**Comment by ahg-g** [comment]
Preemption is an extension point we discussed multiple times in the past, in the last sig meeting we agreed to pursue this once the migration of predicates/priorities is done and settled.

I think adding extensions points to alter more the queue behaviour beyond sorting is certainly valid, but I feel a standalone KEP is warranted.

Regarding adding an Object API to identify tenants, this needs to be discussed with the multi-tenancy working group. I imagine though that having a standalone object that defines a scheduler queue behavior is likely the best path forward (and so a tenant is not defined by a single object, it is a collection of objects: for example a namespace, resource quota in that namespace and for example a SchedulerQueue).

**Comment by yqwang-ms** [comment]
Thanks @ahg-g for the info, totally agree! :)

**Comment by yqwang-ms** [comment]
And this issue can be a driver which helps us to figure out which plugin/featue we should support in future.

**Comment by alculquicondor** [comment]
> 1. Queue sort plugin #87160
> 2. Requeue plugin
>    It is used to judge whether a failed scheduled pod should be requeued now. So that plugin can override the behaiour of unschedulable queue, backoff queue, etc which also impacts on this issue.
>    For example, it should provide a func ShouldRequeueNow(pod) bool.
> 3. Queue empty listener plugin
>    When the active queue is empty, it will call this plugin. This helps to implement the behaiour in my first proposal, i.e. only after it is called, we can requeue the failed scheduled pods, and resnapshot the scheduling view

I would say let's just stay with the Queue endpoint. I believe it alone should suffice to solve starvation. My take would be to start from there. If it isn't enough, a KEP is the way to go.

> 4. Preemption plugin

Let's open a separate issue for this. I also agree we have to engage with multi-tenancy wg before deciding anything.

**Comment by alculquicondor** [comment]
I still would like to understand why you see such starvation with only ~500 unschedulable pods. Given that we should be able to process ~100 pod/s (in 1.17+), we would expect to see any pod scheduled within 5s. Specially if the flush event is the only thing that is bringing pods back into the queue. However, I created #87832, because, if your cluster is static, we still lose backoff information in 60s. Once that's done, for your case, the scheduler should react to new pods immediately. If this is not happening, we might have a bug on what we are already promising.
If you still have other events happening, then #87616 might still reduce starvation for you. Let me know if you can give it a try.

Now, back to extreme cases where there are thousands of unschedulable pods and assuming that the cluster is very static (no nodes being created and no pods finishing, such as the cases described here). All pods would be retried (regardless) every 60s.
So, I started to question why we even do this. The issue that triggered this implementation (#72122) states:
`I don't know if this should be marked as a bug. I don't have any report of such issue. It is more of a precaution to prevent issues.`

So, in your case, it should be safe for you to increase the flush to (say) 10 minutes. Also let me know if you can experiment with this.

But other than configuration parameters, we (kube-scheduler) could also consider this:

- [Extreme] Get rid of the flush, since we don't have evidence that it's actually helping.
- Limit the number of pods we flush back. We would always flush the oldest ones. Either a hard limit (always flush X) or a dynamic limit (stop once the activeQ is past a threshold). This should be low risk and easy to do. WDYT?

**Comment by yqwang-ms** [comment]
Thanks @alculquicondor to look into this and mitigate it :)
Our scheduling result may not match your estimation:
1. The k8s default scheduler is version 1.14
2. We also used scheduler extender to bind/preempt/filter, and it has some very complex logic to select devices and introduced RPC call, which caused the slower scheduling

Increase flush interval and with "your never lose backoff delay features" should help to mitigate my case, but I cannot ensure there are no other cluster events cause pods becomes active again.

**Comment by alculquicondor** [comment]
I created #87850 to discuss these suggestions. However, keep in mind that we probably wouldn't be porting it back.

Sadly, the flush interval (60s) is hardcoded.

**Comment by Huang-Wei** [comment]
@alculquicondor One thing to note is that the 100 pods/s metric is based on fairly simple workloads. It the Pods comes with complex spec such as PodAffinity, a slowdown on scheduling is expected. And as @yqwang-ms , an extender usage would further slowdown a scheduling cycle incredibly (due to marshal/unshaling as well as network overhead).

**Comment by alculquicondor** [comment]
#87948 merged :smiley: 
This should be a second mitigation for the static cluster case. Basically, pods would still get back after 60s, but they would respect the previous backoff.

**Comment by alculquicondor** [comment]
cc @adtac

**Comment by Huang-Wei** [comment]
@yqwang-ms In 1.22 and 1.23, we introduced several enhancements to mitigate this problem. But note they're not applicable for extenders:

1. Fine-grained enqueuing semantics #100347
    
    Plugins can implement an `EventsToRegister()` function to specify which events can move the unschedulable Pods (failed by that plugin) back. In the before, we move Pods back without taking the type of events into consideration.
    In this way, if a Pod failed by plugin X (e.g., missing PVC), only events related to volumes can move this Pod back.

2. Provide an option for plugin developers to move pods to activeQ #103383

    This is useful for expedite the scheduling of a series of Pods that have inter-dependencies.

Moreover, the aforementioned scheduler configuration fields like [PodInitialBackoffSeconds] and [PodMaxBackoffSeconds] can also mitigate the HOL problem.

[PodInitialBackoffSeconds]: https://github.com/kubernetes/kubernetes/blob/a3d62e6925ecbf3be6c39ead50b8e4130a237715/staging/src/k8s.io/kube-scheduler/config/v1beta2/types.go#L81
[PodMaxBackoffSeconds]: https://github.com/kubernetes/kubernetes/blob/a3d62e6925ecbf3be6c39ead50b8e4130a237715/staging/src/k8s.io/kube-scheduler/config/v1beta2/types.go#L86

**Comment by ahg-g** [comment]
@Huang-Wei what would be the best way to verify that #100347 improved things in practice?

**Comment by Huang-Wei** [comment]
We used to talk about dedicated metrics verifying the effect in practice. Before it's taking in place, the quantitative way to verify is to observe `schedule_attempts_total` - the value is expected to be a lot lower than before. In a non-quantitative way, the low-priority pods should have more fair chances to be scheduled.

**Comment by yqwang-ms** [comment]
@Huang-Wei Glad to hear that you have done so much to improve this, thanks! :) Will try to leverage scheduling framework and these plugins.

**Comment by ahg-g** [comment]
@Huang-Wei we have tests running on daily basis on perf-dash, I am wondering if we can add `pod_scheduling_duration_seconds` to [perf-dash](https://github.com/kubernetes/perf-tests/blob/81c96c34e3c1f11c5fe91744ef7ce7bf44c2fe5c/clusterloader2/pkg/measurement/common/scheduler_latency.go#L39), this metric includes the time a pod was waiting until scheduled broken down by attempts, if there is is significant variation across pods, then likely we still have some more work to do.

**Comment by Huang-Wei** [comment]
Are the pods in perf-dash scheduled successfully in one attempt? If so, I don't think `pod_scheduling_duration_seconds` can tell whether #100347 has saved us unnecessary attempts or not.

**Comment by ahg-g** [comment]
> Are the pods in perf-dash scheduled successfully in one attempt? If so, I don't think pod_scheduling_duration_seconds can tell whether #100347 has saved us unnecessary attempts or not.

We don't know, we don't have a metric exposed in the dashboard to track that.

**Comment by alculquicondor** [comment]
/close

Feel free to reopen if you find evidence that the current behavior is not enough.

You might also be interested in #87850 and sig.k8s.io/kueue
